File: app/src/student/java/com.apptronix.nitkonschedule.student/service/DBSyncTask.java
package com.apptronix.nitkonschedule.student.service;

import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;

import com.apptronix.nitkonschedule.model.Course;
import com.apptronix.nitkonschedule.model.CourseResponse;
import com.apptronix.nitkonschedule.model.Schedule;
import com.apptronix.nitkonschedule.model.ScheduleList;
import com.apptronix.nitkonschedule.student.data.DBContract;
import com.apptronix.nitkonschedule.model.Assignment;
import com.apptronix.nitkonschedule.model.AssignmentResponse;
import com.apptronix.nitkonschedule.model.Test;
import com.apptronix.nitkonschedule.model.TestsResponse;
import com.apptronix.nitkonschedule.model.User;
import com.apptronix.nitkonschedule.rest.ApiClient;
import com.apptronix.nitkonschedule.student.rest.ApiInterface;
import com.apptronix.nitkonschedule.student.ui.MainActivity;
import com.apptronix.nitkonschedule.student.widget.TimeTableCollectionWidget;
import com.google.firebase.messaging.FirebaseMessaging;

import org.greenrobot.eventbus.EventBus;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.net.URL;
import java.util.List;
import java.util.Objects;
import java.util.Vector;
import java.util.concurrent.TimeUnit;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Response;
import timber.log.Timber;

/**
 * Created by DevOpsTrends on 7/2/2017.
 */

public class DBSyncTask {

    private static Context mContext;
    public static final MediaType JSON = MediaType.parse("application/json; charset=utf-8");
    static User user;
    static int percentage, attended;

    synchronized public static void syncData(Context context) {

        mContext=context;
        user = new User(mContext);

        fetchAccessToken();

        handleActionFetchCourses();
        handleActionFetchTimeTable();
        handleActionFetchTests();
        handleActionFetchAssignments();
        
        calculatePercentages();
        
        Intent intent = new Intent(context,TimeTableCollectionWidget.class);
        intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
        int[] ids = AppWidgetManager.getInstance(context.getApplicationContext()).getAppWidgetIds(new ComponentName(context.getApplicationContext(),TimeTableCollectionWidget.class));
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS,ids);
        context.sendBroadcast(intent);
        
    }

    private static void calculatePercentages() {
        Cursor coursesCursor = mContext.getContentResolver().query(DBContract.CourseEntry.CONTENT_URI, null, null,null,null);
        if(coursesCursor!=null && coursesCursor.moveToFirst()){
            for(int i=0;i<coursesCursor.getCount();i++){
                coursesCursor.moveToPosition(i);
                String enrolled = coursesCursor.getString(coursesCursor.getColumnIndex(DBContract.CourseEntry.COLUMN_COURSE_ENROLLED));
                if(enrolled!=null && enrolled!="false"){
                    String course = coursesCursor.getString(coursesCursor.getColumnIndex(DBContract.CourseEntry.COLUMN_COURSE));
                    Cursor scheduleCursor = mContext.getContentResolver().query(DBContract.TimeTableEntry.ATTENDANCE_URI,null, DBContract.TimeTableEntry.COLUMN_COURSE+" = ? ",new String[]{course},null);
                    if(scheduleCursor.moveToFirst()){
                        percentage =0;
                        attended=0;
                        for(int j=0;j<scheduleCursor.getCount();j++){
                            scheduleCursor.moveToPosition(j);
                            String present = scheduleCursor.getString(scheduleCursor.getColumnIndex(DBContract.TimeTableEntry.COLUMN_PRESENT));
                            if(present!=null && Objects.equals(present, "false")){
                                attended++;
                            }
                        }
                        percentage=(attended*100)/scheduleCursor.getCount();
                        ContentValues contentValues = new ContentValues();
                        contentValues.put(DBContract.CourseEntry.COLUMN_COURSE,course);
                        contentValues.put(DBContract.CourseEntry.COLUMN_ATT_PERCENT,percentage);
                        mContext.getContentResolver().update(DBContract.CourseEntry.CONTENT_URI,contentValues,
                                DBContract.CourseEntry.COLUMN_COURSE+" = ? ",new String[]{course});
                    }


                }
            }
        }
    }

    private  static void  handleActionFetchCourses() {

        user = new User(mContext);
        ApiInterface apiService = ApiClient.getClient().create(ApiInterface.class);
        Call<CourseResponse> callTT = apiService.getCourses(user.getAccessToken());

        try {
            Response<CourseResponse> response = callTT.execute();
            if(response.isSuccessful()){
                if(response.code()==200){
                    List<Course> courseList = response.body().getResults();
                    insertCourseList(courseList);
                } else if (response.code() == 401) { //bad auth
                    user.setAccessToken(null, mContext); //reset access token
                    fetchAccessToken();
                }
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }




    private static void handleActionFetchTests() {

        Timber.i(user.getAccessToken());
        ApiInterface apiService = ApiClient.getClient().create(ApiInterface.class);
        Call<TestsResponse> callTT = apiService.getTests(user.getAccessToken());

        try {
            Response<TestsResponse> response = callTT.execute();
            if(response.isSuccessful()){
                if(response.code()==200){
                    List<Test> testsList = response.body().getResults();
                    insertTestsList(testsList);
                } else if( response.code() == 401) { //bad auth
                    user.setAccessToken(null, mContext); //reset access token
                    fetchAccessToken();
                }
            }

        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    private static void insertCourseList(List<Course> courseList) {

        Vector<ContentValues> cVVector = new Vector<ContentValues>(courseList.size());

        if (courseList.size() > 0) {

            Timber.i("insert course %d",courseList.size());
            for (Course course : courseList) {
                ContentValues courseCV = new ContentValues();
                courseCV.put(DBContract.CourseEntry.COLUMN_COURSE, course.getName());
                courseCV.put(DBContract.CourseEntry.COLUMN_COURSE_DESCRIPTION, course.getDescription());
                cVVector.add(courseCV);

                FirebaseMessaging.getInstance().subscribeToTopic(course.getName());
            }
            ContentValues[] cvArray = new ContentValues[cVVector.size()];
            cVVector.toArray(cvArray);
            int inserts = mContext.getContentResolver().bulkInsert(DBContract.CourseEntry.CONTENT_URI, cvArray);
            Timber.i("%d course bulk insert success", inserts);
        }
    }

    private static void handleActionFetchAssignments() {

        ApiInterface apiService = ApiClient.getClient().create(ApiInterface.class);
        Call<AssignmentResponse> callTT = apiService.getAssignments(user.getAccessToken());

        try {
            Response<AssignmentResponse> response = callTT.execute();
            if(response.isSuccessful()){
                if(response.code()==200){
                    List<Assignment> assgnList = response.body().getResults();
                    insertAssignmentsList(assgnList);
                } else if( response.code() == 401) { //bad auth
                    user.setAccessToken(null, mContext); //reset access token
                    fetchAccessToken();
                }
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static void fetchAccessToken() {

        User user = new User(mContext);
        String refreshToken = user.getRefreshToken();
        Timber.i("sending refresh token %s to server",refreshToken);

        try {

            OkHttpClient.Builder builder = new OkHttpClient.Builder();
            builder.connectTimeout(30, TimeUnit.SECONDS);
            builder.readTimeout(30, TimeUnit.SECONDS);
            builder.writeTimeout(30, TimeUnit.SECONDS);
            OkHttpClient client = builder.build();

            JSONObject postJSON = new JSONObject();
            postJSON.put("refreshToken",refreshToken);
            RequestBody body = RequestBody.create(JSON, String.valueOf(postJSON));
            Request request = new Request.Builder()
                    .url(new URL(ApiClient.BASE_URL+"getAccessToken"))
                    .post(body)
                    .build();


            okhttp3.Response response = client.newCall(request).execute();

            if(response.isSuccessful()){

                String responseString = response.body().string();
                Timber.i("AccessToken Response is %s",responseString);

                if(responseString.equals("fail")){

                    EventBus.getDefault().post(new MainActivity.MessageEvent("TokenUpdateRefused"));

                } else {

                    //registration, received refresh token
                    user.updateTokens(user.getRefreshToken(),responseString,mContext);

                }

            } else {

                EventBus.getDefault().post(new MainActivity.MessageEvent("ServerUnreachable"));

            }

        } catch (IOException | JSONException e1) {
            e1.printStackTrace();
            EventBus.getDefault().post(new MainActivity.MessageEvent("ServerUnreachable"));
        }
    }

    private static void handleActionFetchTimeTable(){

        user = new User(mContext);
        ApiInterface apiService = ApiClient.getClient().create(ApiInterface.class);
        Call<ScheduleList> callTT = apiService.getSchedule(user.getAccessToken());
        Timber.i("user AT %s",user.getAccessToken());
        try {
            Response<ScheduleList> response = callTT.execute();
            if(response.isSuccessful()){
                if(response.code()==200){
                    List<Schedule> timeTableList = response.body().getResults();
                    insertTimeTableList(timeTableList);
                } else if( response.code() == 401) { //bad auth
                    user.setAccessToken(null, mContext); //reset access token
                    fetchAccessToken();
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static void insertTestsList(List<Test> testsList) {

        Vector<ContentValues> cVVector = new Vector<ContentValues>(testsList.size());

        for(Test test:testsList){
            ContentValues testCV = new ContentValues();
            testCV.put(DBContract.TestsEntry.COLUMN_COURSE_CODE,test.getCourse());
            testCV.put(DBContract.TestsEntry.COLUMN_TITLE,test.getTitle());
            testCV.put(DBContract.TestsEntry.COLUMN_SYLLABUS,test.getPortions());
            testCV.put(DBContract.TestsEntry.COLUMN_TEST_DATE,test.getTestDate());
            testCV.put(DBContract.TestsEntry.COLUMN_WEIGHTAGE,test.getWeightage());
            cVVector.add(testCV);
        }

        if ( cVVector.size() > 0 ) {
            ContentValues[] cvArray = new ContentValues[cVVector.size()];
            cVVector.toArray(cvArray);
            int inserts=mContext.getContentResolver().bulkInsert(DBContract.TestsEntry.CONTENT_URI, cvArray);
            Timber.i("%d tests bulk insert success", inserts);
        }

    }

    private static void insertAssignmentsList(List<Assignment> assgnList) {

        Vector<ContentValues> cVVector = new Vector<ContentValues>(assgnList.size());

        for(Assignment assgn:assgnList){
            ContentValues assgnCV = new ContentValues();
            assgnCV.put(DBContract.AssignmentsEntry.COLUMN_COURSE_CODE,assgn.getCourse());
            assgnCV.put(DBContract.AssignmentsEntry.COLUMN_TITLE,assgn.getTitle());
            assgnCV.put(DBContract.AssignmentsEntry.COLUMN_DESCRIPTION,assgn.getDescription());
            assgnCV.put(DBContract.AssignmentsEntry.COLUMN_SUBMISSION_DATE,assgn.getSubmissionDate());
            assgnCV.put(DBContract.AssignmentsEntry.COLUMN_WEIGHTAGE,assgn.getWeightage());
            cVVector.add(assgnCV);
        }

        if ( cVVector.size() > 0 ) {
            ContentValues[] cvArray = new ContentValues[cVVector.size()];
            cVVector.toArray(cvArray);
            int inserts=mContext.getContentResolver().bulkInsert(DBContract.AssignmentsEntry.CONTENT_URI, cvArray);
            Timber.i("%d assignments bulk insert success", inserts);
        }

    }


    private static void insertTimeTableList(List<Schedule> timeTableList) {

        Vector<ContentValues> cVVector = new Vector<ContentValues>(timeTableList.size()*8);

        for(Schedule dayTable:timeTableList){

            ContentValues timeTableCV = new ContentValues();
            timeTableCV.put(DBContract.TimeTableEntry.COLUMN_DATE,dayTable.getDate());
            timeTableCV.put(DBContract.TimeTableEntry.COLUMN_COURSE,dayTable.getCourse());
            timeTableCV.put(DBContract.TimeTableEntry.COLUMN_TIME,dayTable.getTime());
            timeTableCV.put(DBContract.TimeTableEntry.COLUMN_DESCRIPTION,dayTable.getDescription());
            timeTableCV.put(DBContract.TimeTableEntry.COLUMN_PRESENT,dayTable.getPresentIDs());
            cVVector.add(timeTableCV);


        }

        if ( cVVector.size() > 0 ) {
            ContentValues[] cvArray = new ContentValues[cVVector.size()];
            cVVector.toArray(cvArray);
            int inserts=mContext.getContentResolver().bulkInsert(DBContract.TimeTableEntry.CONTENT_URI, cvArray);
            Timber.i("%d day's time table bulk insert success", inserts);
        }
    }

}

File: app/src/teacher/java/com/apptronix/nitkonschedule/teacher/service/DBSyncTask.kt
package com.apptronix.nitkonschedule.teacher.service

import android.annotation.SuppressLint
import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.ContentValues
import android.content.Context
import android.content.Intent
import com.apptronix.nitkonschedule.teacher.model.*
import com.apptronix.nitkonschedule.rest.ApiClient
import com.apptronix.nitkonschedule.teacher.data.DBContract
import com.apptronix.nitkonschedule.teacher.rest.ApiInterface
import com.apptronix.nitkonschedule.teacher.ui.MainActivity
import com.apptronix.nitkonschedule.teacher.ui.fragments.ScheduleFragment
import com.apptronix.nitkonschedule.teacher.widget.TimeTableCollectionWidget
import com.google.firebase.messaging.FirebaseMessaging
import okhttp3.MediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import org.greenrobot.eventbus.EventBus
import org.json.JSONException
import org.json.JSONObject
import timber.log.Timber
import java.io.IOException
import java.net.URL
import java.util.*
import java.util.concurrent.TimeUnit

@SuppressLint("StaticFieldLeak")
/**
 * Created by DevOpsTrends on 7/2/2017.
 */

object DBSyncTask {

    private var mContext: Context? = null
    val JSON = MediaType.parse("application/json; charset=utf-8")
    private var user: User? = null

    @Synchronized
    fun syncData(context: Context) {
        mContext = context

        user=User(context)
        fetchAccessToken(true)
        handleActionFetchTests()
        handleActionFetchCourses()
        handleActionFetchTimeTable()
        handleActionFetchAssignments()
        handleActionFetchResources()

        val intent = Intent(context, TimeTableCollectionWidget::class.java)
        intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
        val ids = AppWidgetManager.getInstance(context.applicationContext).getAppWidgetIds(ComponentName(context.applicationContext, TimeTableCollectionWidget::class.java))
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
        context.sendBroadcast(intent)
    }

    private fun handleActionFetchTests() {

        Timber.i(user!!.accessToken)
        val apiService = ApiClient.getClient().create(ApiInterface::class.java)
        val callTT = apiService.getTests(user!!.accessToken)

        try {
            val response = callTT.execute()
            if (response.isSuccessful) {
                if (response.code() == 200) {
                    val testsList = response.body()!!.results
                    insertTestsList(testsList)
                } else if (response.code() == 401|| response.code() == 500) { //bad auth
                    user!!.setAccessToken(null, mContext!!) //reset access token
                    fetchAccessToken(true)
                }
            }

        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    private fun handleActionFetchAssignments() {

        val apiService = ApiClient.getClient().create(ApiInterface::class.java)
        val callTT = apiService.getAssignments(user!!.accessToken)

        try {
            val response = callTT.execute()
            if (response.isSuccessful) {
                if (response.code() == 200) {
                    val assgnList = response.body()!!.results
                    insertAssignmentsList(assgnList)
                } else if (response.code() == 401|| response.code() == 500) { //bad auth
                    user!!.setAccessToken(null, mContext!!) //reset access token
                    fetchAccessToken(true)
                }
            }

        } catch (e: IOException) {
            e.printStackTrace()
        }

    }


    private fun handleActionFetchResources() {

        val apiService = ApiClient.getClient().create(ApiInterface::class.java)
        val callTT = apiService.getResources(user!!.accessToken)

        try {
            val response = callTT.execute()
            if (response.isSuccessful) {
                if (response.code() == 200) {
                    val resourcesList = response.body()!!.results
                    insertResourcesList(resourcesList)
                } else if (response.code() == 401|| response.code() == 500) { //bad auth
                    user!!.setAccessToken(null, mContext!!) //reset access token
                    fetchAccessToken(true)
                }
            }

        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    private fun handleActionFetchCourses() {

        val apiService = ApiClient.getClient().create(ApiInterface::class.java)
        val callTT = apiService.getCourses(user!!.accessToken)

        try {
            val response = callTT.execute()
            if (response.isSuccessful) {
                Timber.i(response.body()!!.results.toString())
                if (response.code() == 200 && response.body()!!.results != null) {
                    val courseList = response.body()!!.results
                    insertCourseList(courseList)
                } else if (response.code() == 401|| response.code() == 500) { //bad auth
                    user!!.setAccessToken(null, mContext!!) //reset access token
                    fetchAccessToken(true)
                }
            }

        } catch (e: IOException) {
            e.printStackTrace()
        }


    }

    private fun fetchAccessToken(force:Boolean) {


        if(!force){
            if(!(user!!.accessTokenTime == null || 0.equals(user!!.accessTokenTime))){
                if((user!!.accessTokenTime!! +300)>(System.currentTimeMillis()/1000)) { //not expired yet
                    return
                }
            }

        }

        val refreshToken = user!!.refreshToken
        Timber.i("sending refresh token %s to server", refreshToken)

        try {

            val builder = OkHttpClient.Builder()
            builder.connectTimeout(30, TimeUnit.SECONDS)
            builder.readTimeout(30, TimeUnit.SECONDS)
            builder.writeTimeout(30, TimeUnit.SECONDS)
            val client = builder.build()

            val postJSON = JSONObject()
            postJSON.put("refreshToken", refreshToken)
            val body = RequestBody.create(JSON, postJSON.toString())
            val request = Request.Builder()
                    .url(URL(ApiClient.BASE_URL + "getAccessToken"))
                    .post(body)
                    .build()


            val response = client.newCall(request).execute()

            if (response.isSuccessful) {

                val responseString = response.body()!!.string()
                Timber.i("AccessToken Response is %s", responseString)

                if (responseString == "fail") {

                    EventBus.getDefault().post(MainActivity.MessageEvent("TokenUpdateRefused"))

                } else {

                    //registration, received refresh token
                    user!!.updateTokens(user!!.refreshToken, responseString, mContext!!)

                }

            } else {

                EventBus.getDefault().post(MainActivity.MessageEvent("ServerUnreachable"))

            }

        } catch (e1: IOException) {
            e1.printStackTrace()
            EventBus.getDefault().post(MainActivity.MessageEvent("ServerUnreachable"))
        } catch (e1: JSONException) {
            e1.printStackTrace()
            EventBus.getDefault().post(MainActivity.MessageEvent("ServerUnreachable"))
        }

    }

    private fun handleActionFetchTimeTable() {

        val apiService = ApiClient.getClient().create(ApiInterface::class.java)
        val callTT = apiService.getSchedule(user!!.accessToken)

        try {
            val response = callTT.execute()
            if (response.isSuccessful) {
                if (response.code() == 200) {
                    val timeTableList = response.body()!!.results
                    insertTimeTableList(timeTableList)
                } else if (response.code() == 401|| response.code() == 500) { //bad auth
                    user!!.setAccessToken(null, mContext!!) //reset access token
                    fetchAccessToken(true)
                }
            }

        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    private fun insertTestsList(testsList: List<Test>?) {

        val cVVector = Vector<ContentValues>(testsList!!.size)

        for (test in testsList) {
            val assgnCV = ContentValues()
            assgnCV.put(DBContract.TestsEntry.COLUMN_COURSE_CODE, test.course)
            assgnCV.put(DBContract.TestsEntry.COLUMN_TITLE, test.title)
            assgnCV.put(DBContract.TestsEntry.COLUMN_SYLLABUS, test.portions)
            assgnCV.put(DBContract.TestsEntry.COLUMN_TEST_DATE, test.testDate)
            assgnCV.put(DBContract.TestsEntry.COLUMN_WEIGHTAGE, test.weightage)
            cVVector.add(assgnCV)
        }

        if (cVVector.size > 0) {
            val cvArray = cVVector.toTypedArray()
            val inserts = mContext!!.contentResolver.bulkInsert(DBContract.TestsEntry.CONTENT_URI, cvArray)
            Timber.i("%d tests bulk insert success", inserts)
        }

    }


    private fun insertAssignmentsList(assgnList: List<Assignment>?) {

        val cVVector = Vector<ContentValues>(assgnList!!.size)

        for (assgn in assgnList) {
            val assgnCV = ContentValues()
            assgnCV.put(DBContract.AssignmentsEntry.COLUMN_COURSE_CODE, assgn.course)
            assgnCV.put(DBContract.AssignmentsEntry.COLUMN_TITLE, assgn.title)
            assgnCV.put(DBContract.AssignmentsEntry.COLUMN_DESCRIPTION, assgn.description)
            assgnCV.put(DBContract.AssignmentsEntry.COLUMN_SUBMISSION_DATE, assgn.submissionDate)
            assgnCV.put(DBContract.AssignmentsEntry.COLUMN_WEIGHTAGE, assgn.weightage)
            cVVector.add(assgnCV)
        }

        if (cVVector.size > 0) {
            val cvArray = cVVector.toTypedArray()
            val inserts = mContext!!.contentResolver.bulkInsert(DBContract.AssignmentsEntry.CONTENT_URI, cvArray)
            Timber.i("%d assignments bulk insert success", inserts)
        }

    }

    private fun insertCourseList(courseList: List<Course>?) {

        val cVVector = Vector<ContentValues>(courseList!!.size)

        if (courseList.size > 0) {

            Timber.i("insert course %d",courseList.size)
            for (course in courseList) {
                val courseCV = ContentValues()
                var regIDs :String?=""
                for(id in course.enrolled){
                    regIDs+=id+";"
                }
                courseCV.put(DBContract.CourseEntry.COLUMN_COURSE, course.name)
                courseCV.put(DBContract.CourseEntry.COLUMN_COURSE_ENROLLED_IDS, regIDs)
                courseCV.put(DBContract.CourseEntry.COLUMN_COURSE_DESCRIPTION, course.description)
                cVVector.add(courseCV)

                FirebaseMessaging.getInstance().subscribeToTopic(course.name)
            }
            val cvArray = cVVector.toTypedArray()
            val inserts = mContext!!.contentResolver.bulkInsert(DBContract.CourseEntry.CONTENT_URI, cvArray)
            Timber.i("%d course bulk insert success", inserts)
        }
    }

    private fun insertTimeTableList(timeTableList: List<Schedule>?) {

        val cVVector = Vector<ContentValues>(timeTableList!!.size)

        for (dayTable in timeTableList) {

            val timeTableCV = ContentValues()
            timeTableCV.put(DBContract.TimeTableEntry.COLUMN_DATE, dayTable.getDate())
            timeTableCV.put(DBContract.TimeTableEntry.COLUMN_COURSE, dayTable.course)
            timeTableCV.put(DBContract.TimeTableEntry.COLUMN_TIME, dayTable.time)
            timeTableCV.put(DBContract.TimeTableEntry.COLUMN_DESCRIPTION, dayTable.description)
            timeTableCV.put(DBContract.TimeTableEntry.COLUMN_PRESENT_IDS, dayTable.presentIDs)

            Timber.i("%s %d",dayTable.description,dayTable.date)
            cVVector.add(timeTableCV)

        }

        if (cVVector.size > 0) {
            val cvArray = cVVector.toTypedArray()
            val inserts = mContext!!.contentResolver.bulkInsert(DBContract.TimeTableEntry.CONTENT_SCHEDULE_URI, cvArray)
            Timber.i("%d day's time table bulk insert success", inserts)
            EventBus.getDefault().post(ScheduleFragment.MessageEvent("reload"))
        }
    }

    private fun insertResourcesList(resourceList: List<ResourceModel>?) {

        val cVVector = Vector<ContentValues>(resourceList!!.size)

        for (res in resourceList) {

            for (file in res.resources) {
                val resourceCV = ContentValues()
                resourceCV.put(DBContract.RepositoryEntry.COLUMN_FILE_NAME, file)
                resourceCV.put(DBContract.RepositoryEntry.COLUMN_COURSE, res.course)

                cVVector.add(resourceCV)
            }


        }

        if (cVVector.size > 0) {
            val cvArray = cVVector.toTypedArray()
            val inserts = mContext!!.contentResolver.bulkInsert(DBContract.RepositoryEntry.CONTENT_URI, cvArray)
            Timber.i("%d resource insert success", inserts)
        }
    }

}
